pokemon-service/src/main/java/cz/muni/fi/pa165/pokemon/league/participation/manager/service/GymServiceImpl.java

```diff
@@ -1,126 +1,151 @@
+import cz.muni.fi.pa165.pokemon.league.participation.manager.exceptions.EntityIsUsedException;
 import java.util.List;
+import java.util.NoSuchElementException;
 import javax.inject.Inject;
 import java.util.stream.Collectors;
 import org.springframework.dao.DataAccessException;
 
 /**
  * Implementation of GymService.
- * 
+ *
  * @author Tamás Rózsa 445653
  */
 public class GymServiceImpl implements GymService {
 
     @Inject
     private GymDAO gymDAO;
-    
+
+    @Inject
+    private BadgeDAO badgeDAO;
+
     @Override
-    public void createGym(Gym gym) {
+    public void createGym(Gym gym) throws EntityIsUsedException {
+        Gym usedGym = this.findGymByLeader(gym.getGymLeader());
+        if (usedGym != null) {
+            throw new EntityIsUsedException("Trainer already used in gym " + usedGym.toString());
+        }
         try {
             gymDAO.createGym(gym);
-        }
-        catch(Exception ex) {
-            throw new DataAccessException("Cannot create new gym " + gym.toString(), ex) {};
+        } catch (Exception ex) {
+            throw new DataAccessException("Cannot create new gym " + gym.toString(), ex) {
+            };
         }
     }
 
     @Override
-    public void updateGymLocation(Gym gym, String newLocaton) {
+    public void updateGymLocation(Gym gym, String newLocation) {
         try {
-            gym.setLocation(newLocaton);
+            gym.setLocation(newLocation);
             gymDAO.updateGym(gym);
-        }
-        catch(Exception ex) {
-            throw new DataAccessException("Cannot update location fot gym " + gym.toString(), ex) {};
+        } catch (Exception ex) {
+            throw new DataAccessException("Cannot update location fot gym " + gym.toString(), ex) {
+            };
         }
     }
 
     @Override
-    public void changeGymType(Gym gym, Trainer trainer, PokemonType newType) {
+    public boolean changeGymType(Gym gym, Trainer trainer, PokemonType newType) {
         try {
             if (gym.getGymLeader().equals(trainer)) {
                 gym.setType(newType);
                 gymDAO.updateGym(gym);
+                return true;
             }
-        }
-        catch(Exception ex) {
-            throw new DataAccessException("Cannot change gym type for gym " + gym.toString(), ex) {};
+            return false;
+        } catch (Exception ex) {
+            throw new DataAccessException("Cannot change gym type for gym " + gym.toString(), ex) {
+            };
         }
     }
 
     @Override
-    public void changeGymLeader(Gym gym, Trainer newGymLeader) {
+    public void changeGymLeader(Gym gym, Trainer newGymLeader) throws EntityIsUsedException {
+        if (!newGymLeader.equals(gym.getGymLeader())) {
+            Gym usedGym = this.findGymByLeader(newGymLeader);
+            if (usedGym != null) {
+                throw new EntityIsUsedException("Gymleader already use on Gym " + usedGym.toString());
```

Comment:
Looking at the Javadoc of the exception, this isn't the right one to use, however, since its name fits, I'm okay with it, provided that the Javadoc is updated.